torch/_functorch/_aot_autograd/schemas.py

```diff
@@ -384,19 +339,10 @@ def __post_init__(self):
         # separately.
         self.num_outputs_rng_offset = 1 if self.is_rng_op_functionalized else 0
 
-        # Our forward() returns both (mutated_inputs, outputs, output_intermediate_bases, saved_tensors, saved_symints)
-        self.num_forward_returns = (
-            self.num_mutated_inp_runtime_indices
-            + self.num_outputs
-            + self.num_intermediate_bases
-        )
-        # In case of functionalization of rng ops, the fw_module returns one
-        # additional output for rng offset. This rng offset is used right
-        # away to advance the rng state, and is not passed on to the raw
-        # outputs. However, we need to know the exact boundary to identify
-        # which tensors to be saved for the bwd graph.  num_forward captures
-        # this information.
-        self.num_forward = self.num_forward_returns + self.num_outputs_rng_offset
+        # Many parameters dependon requires_subclass_dispatch, and for convenience we
+        # want to be able to modify requires_subclass_dispatch on an already-constructed
+        # ViewAndMutationMeta.
+        self.set_requires_subclass_dispatch(self.requires_subclass_dispatch)
```

Comment:
A bit hard to tell - is everything in this function just code motion from what you removed above, other than the `self.requires_subclass_dispatch = requires_subclass_dispatch` field?

I think it's a bit confusing that this method (which seems only related to subclasses) sets a bunch of additional metadata that is used in non-subclass cases. Any reason for that change?

Reply:
Yes, this just code motion. Everything that I moved depends on the value of `requires_subclass_dispatch`.

Alternatively I could leave it in `__post_init__`; and then when I set requires_subclass_dispatch, I manually call `__post_init__()`? Or open to whatever other suggestions - I'm not really sure what's the right pattern to use here

Reply:
Oh right, I see: it's because the logic that computes the `mutated_inp_runtime_indices` field depends on this flag downstream.

Let me try to summarize the situation, since I think it is pretty not-ideal:

(1) We have some logic that decides, for every input mutation, whether it can go in the graph or outside the graph

(2) @tugsbayasgalan added support for "keeping input mutations in the graph when they don't require grad, even if they are part of a training graph": https://github.com/pytorch/pytorch/pull/111046. In that PR, we ended up with the following split, which looks roughly like:

(2a) In general, the decision is roughly "if the input requires grad and was mutated, keep the mutation outside the graph, otherwise put it in the graph"

(2b) However, this posed some challenges subclass handling, so the logic became "BUT if any subclasses are involved, then just don't put any input mutations in the graph".

I think the bug from the current PR shows that this strategy is brittle. So ideally, I think it would be cleaner for the fix to be "treat input mutations the same way regardless of whether or not subclasses are involved".

cc @tugsbayasgalan what were the difficulties with doing this the first time around?